docs_src/design/ucr/0005-Record-and-Replay.md

```diff
@@ -0,0 +1,48 @@
+## Record and Replay
+### Submitters
+- Lenny Goodell (Intel)
+- Jim White (IOTech)
+
+## Change Log
+- [pending](URL of PR) (2022-08-11)
+
+### Market Segments
+- Any/All
+
+### Motivation
+Currently one must have physical devices and appropriate environment to produce real device data (Event/Readings) into an EdgeX solution for other EdgeX services (Core Data, App Services, eKuiper Rules Engine) to consume. This is often not the case when someone is developing/testing one of these consuming EdgeX services. A good example of this is the **[RFID LLRP Inventory App Service](https://github.com/edgexfoundry/app-rfid-llrp-inventory)** . In order to test this service it is dependent on the [RFID LLRP Device Service](https://github.com/edgexfoundry/device-rfid-llrp-go), physical LLRP RFID readers, RFID Tags and environment where these are deployed. Having a way to record EdgeX the Event/Readings from an actual deployment that then could be replayed in development environment for testing would be very valuable. 
+
+Other potential uses are:
+
+- “Replayable data” could be used to validate a new or updated EdgeX service behaves like the existing one (EdgeX certification). Or the data could be offered by component providers to give users an example of the type of data generated from one of their device services.
+- EdgeX Ready applicants could use a record feature to capture data from their device service and the committee could explore that in relation to their profile to make sure they seemed to know what they were doing.
+- Record and replay can be used to reproduce a reported bug.
+- Record and replay could serve as a means to backup data in the event of data loss.
+
+### Target Users
+- Device Manufacturer
+- Device Owner
+- Device User
+- Device Maintainer
+- Software Developer
+- Software Deployer
+- Software Integrator
+
+### Description
+Target users have the need to be able to replay recoded EdgeX Event/Reading for functional, performance or reproducible testing. This UCR describes a new capability that allows user to first record Event/Readings from real devices in real-time and than be able to replay the Event/Readings as if it was in real-time.
+
+### Existing solutions
+There are simulators for some devices (i.e. Modbus), but there isn't an general solution to reproduce real device data into EdgeX without the physical devices being present.
+
+### Requirements
+- Record device Event/Reading data (at rates, volumes and details like timestamps) 
+- Replay captured data at captured, slower or faster speeds
+- Export captured data for use at a later time or to send to other users
+- Import recoded data that was previously exported
+- Replay recorded data from multiple sources at the same time (this mimics more (device) services feeding EdgeX )
+
```

Comment:
```suggestion
- Option to record all resource values from a service, including the latest values at the time of capture, for resources that were not expressed in events during the capture.  (So to include the static data, not just the dynamic data.)
```

Reply:
The static data (resources and device definitions) are already captured in the Device Profile and Device files. IMO, we don't need to provide a mechanism to export these since they are readily available via the files.

Reply:
That would be OK for purely static data, but some things change only infrequently and, to capture the true state of the system, you must also have their values. Rather than obsess over whether something is truly static or "mostly static", I'm suggesting capturing the last known value of every Device Resource of every Device covered by the capture.
This is similar to why I wanted the last known value of every resource to persist in core-data.

Reply:
OK, so capture devices and device profiles reference in the captured data?
The replay is more complicated as it now must override the current profiles and devices with that captured. 
IMO, this goes beyond the scope of _**replay recorded EdgeX Event/Readings for functional, performance or reproducible testing**_. In this scope that data is static.

Reply:
Actually, for device service that do auto provisioning, we do need to capture the Device definitions for the devices refenced in the data, since this data is not static, i.e. now device definition file. Still don't see the need for exporting the device profile, but at this point it is just another piece of data... ;-) @jpwhitemn , thoughts?

Reply:
Added requirements for this